README.md

```diff
@@ -155,10 +155,9 @@ The TEAL assembly smart contract language uses program branches with no loops (i
-| ["set admin" "transfer admin"](bin/set-transfer-admin.sh) | Gives an account transfer admin rights. Transfer admins are responsible for managing who can transfer and when. | contract admin  |
-| ["transfer restrictions"](tests/set_transfer_restrictions.test.js) | Sets account transfer restrictions: 1) `freeze` – freezes a specific address. 2) `max balance` – sets the max number of tokens an account can hold. 3) `lock until` – stop transfers from the address until the specified date. A locked address can still receive tokens but it cannot send them until the lockup time. 4) `transfer group` –  sets the category of an address for use in transfer group rules. The default category is 1. | transfer admin |
-| ["transfer group" "lock"](bin/transfer-group-lock.sh) | Specifies a lock until time for transfers between a transfer from-group and a to-group. Transfers can between groups can only occur after the lock until time. The lock until time is specified as a Unix timestamp integer in seconds since the Unix Epoch. By default transfers beetween groups are not allowed. To allow a transfer set a timestamp in the past such as "1" - for the from and to group pair . The special transfer group default number "0" means the transfer is blocked.  | transfer admin |
-| ["mint"](bin/mint.sh) | Create new tokens from the reserve | contract admin |
-| ["burn"](tests/burn.test.js) | Destroy tokens from a specified address | contract admin |
+| ["set permissions"](tests/permissions.test.js) | Sets account contract permissions. Accepts 4-bit permissions integer, where each bit represents a role: 1) account transfer restrictions (wallets admin), 2) transfer rules (transfer rules admin), 3) mint/burn (assets admin), 4) granting permissions (contract admin) | contract admin |
+| ["transfer restrictions"](tests/set_transfer_restrictions.test.js) | Sets account transfer restrictions: 1) `freeze` – freezes a specific address. 2) `max balance` – sets the max number of tokens an account can hold. 3) `lock until` – stop transfers from the address until the specified date. A locked address can still receive tokens but it cannot send them until the lockup time. 4) `transfer group` –  sets the category of an address for use in transfer group rules. The default category is 1. | wallets admin |
+| ["transfer group" "lock"](bin/transfer-group-lock.sh) | Specifies a lock until time for transfers between a transfer from-group and a to-group. Transfers can between groups can only occur after the lock until time. The lock until time is specified as a Unix timestamp integer in seconds since the Unix Epoch. By default transfers beetween groups are not allowed. To allow a transfer set a timestamp in the past such as "1" - for the from and to group pair . The special transfer group default number "0" means the transfer is blocked. | transfer rules admin |
+| ["mint"](bin/mint.sh) | Create new tokens from the reserve | assets admin |
```

Comment:
Need a chart in the README that shows how the permissions flag integers map to permissions. This is currently shown in the PyTeal but needs to also be in the README.

Reply:
@Alexey1100 I was just thinking through this a bit more. I think this is too big of a matrix. It's kind of satisfying but I think people are going to get overwhelmed with the complexity and make mistakes.

| Role Integer | Admin Roles                                           | Bit Mask Representation | Contract Admin | Mint/Burn | Transfer Rules | Wallets |
| ------------ | ----------------------------------------------------- | ----------------------- | -------------- | --------- | -------------- | ------- |
| **0**        | None (Default)                                        | 0000                    | 0              | 0         | 0              | 0       |
| **1**        | Wallets                                               | 0001                    | 0              | 0         | 0              | 1       |
| **2**        | Transfer Rules                                        | 0010                    | 0              | 0         | 1              | 0       |
| 3            | Transfer Rules + Wallets                              | 0011                    | 0              | 0         | 1              | 1       |
| **4**        | Mint/Burn                                             | 0100                    | 0              | 1         | 0              | 0       |
| _5_          | Mint/Burn + Wallets                                   | 0101                    | 0              | 1         | 0              | 1       |
| _6_          | Mint/Burn + Transfer Rules                            | 0110                    | 0              | 1         | 1              | 0       |
| _7_          | Mint/Burn + Transfer Rules + Wallets                  | 0111                    | 0              | 1         | 1              | 1       |
| **8**        | Contract Admin                                        | 1000                    | 1              | 0         | 0              | 0       |
| _9_          | Contract Admin + Wallets                              | 1001                    | 1              | 0         | 0              | 1       |
| 10           | Contract Admin + Transfer Rules                       | 1010                    | 1              | 0         | 1              | 0       |
| _11_         | Contract Admin + Transfer Rules + Wallets             | 1011                    | 1              | 0         | 1              | 1       |
| _12_         | Contract Admin + Mint/Burn                            | 1100                    | 1              | 1         | 0              | 0       |
| _13_         | Contract Admin + Mint/Burn + Wallets                  | 1101                    | 1              | 1         | 0              | 1       |
| _14_         | Contract Admin + Mint/Burn + Transfer Rules           | 1110                    | 1              | 1         | 1              | 0       |
| _15_         | Contract Admin + Mint/Burn + Transfer Rules + Wallets | 1111                    | 1              | 1         | 1              | 1       |

## Simpler Approach

### Plan A

I would just check if there's enough memory available to have a separate int for each role. If there is I think it will be much more legible. Also, it will be easier for folks to read the global explorer data to understand which account has what role.

### Plan B

If not we can't afford 4 ints we can partition the roles into two bitmasks with 4 variations:
*  the super user 2 bit mask matrix with Contract Admin & Mint/Burn roles.
* the lower powered 2 bit mask matrix with Transfer Rules + Wallet roles.

### Plan C

If there's only room for the 1 int we can use the full 4 bit mask.

Whichever we choose we should write unit tests for each of the integers available in the 1 bit (with 4 separate roles), 2 bit or 4 bit.

Reply:
@aquabu we are quite close to the limit right now with 966 bytes contract size, but I can check whether we can implement plan B or even A :)

I personally don't think the current permission system might be overwhelming, since it's a standard UNIX approach to permissions and developers are used to it.

Regarding the readability, I don't think anyone would be reading or writing this values manually without any tests on production, and code imlementation of it is trivial in any environment.

Reply:
@aquabu I agree we should add an additional integration test which will go through all combination of permission bits and codepaths.

Even though we would be basically just testing the TEAL Bitwise AND function, it will make us feel safer :)